internal/bundle.go

```diff
@@ -404,81 +402,126 @@ func (bundle *Bundle) getFileUpdateCount(filePath string) uint64 {
 		// and assign the update count from corresponding tables
 		return 0
 	}
-	fileStat, ok := bundle.TableStatistics[*relFileNode]
+	fileStat, ok := bundle.RelationsStats[*relFileNode]
 	if !ok {
 		return 0
 	}
-	return fileStat.nTupleDeleted + fileStat.nTupleUpdated + fileStat.nTupleInserted
+	return fileStat.deletedTuplesCount + fileStat.updatedTuplesCount + fileStat.insertedTuplesCount
 }
 
 func (bundle *Bundle) PackTarballs() (map[string][]string, error) {
-	headers, fileInfos := bundle.TarBallComposer.Compose()
-	// write headers to separate tarball
-	err := bundle.writeHeaders(headers)
+	headers, tarFilesCollections := bundle.TarBallComposer.Compose()
+	headersTarName, headersNames, err := bundle.writeHeaders(headers)
 	if err != nil {
 		return nil, err
 	}
 	tarFileSets := make(map[string][]string, 0)
-	// todo: rewrite bundle queue and support parallel tarballs writing
-	tarBall := bundle.Deque()
-	tarBall.SetUp(bundle.Crypter)
-	previousUpdateRating := uint64(0)
-	for _, fileInfo := range fileInfos {
-		tarBall = bundle.CheckTarBall(tarBall, previousUpdateRating, fileInfo.updateRating)
-		previousUpdateRating = fileInfo.updateRating
-		err := bundle.packFileIntoTar(fileInfo, tarBall)
-		if err != nil {
-			return nil, err
+	tarFileSets[headersTarName] = headersNames
+
+	for _, tarFilesCollection := range tarFilesCollections {
+		tarBall := bundle.Deque()
+		tarBall.SetUp(bundle.Crypter)
+		for _, composeFileInfo := range tarFilesCollection.files {
+			tarFileSets[tarBall.Name()] = append(tarFileSets[tarBall.Name()], composeFileInfo.header.Name)
 		}
-		tarFileSets[tarBall.Name()] = append(tarFileSets[tarBall.Name()], fileInfo.header.Name)
+		// tarFilesCollection closure
+		tarFilesCollectionLocal := tarFilesCollection
+		go func() {
+			for _, fileInfo := range tarFilesCollectionLocal.files {
+				err := bundle.packFileIntoTar(fileInfo, tarBall)
+				if err != nil {
+					panic(err)
```

Comment:
- panic here will cancel creating backup without any fallback
- panic should be used rarely, only in cases where we can't handle it and it really looks like undefined behavior
- before modification, there was an error handling, so I guess it still should be error handling
- read about ErrorGroup https://godoc.org/golang.org/x/sync/errgroup, It allows to cancel all if any throw an error